packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,562 @@
+
+  @override
+  String toString() {
+    return 'MaterialGap(key: $key, child: $size)';
+  }
+}
+
+/// A widget useful for animating the merging and separating of [Material]
+/// slices. The widget takes [MergeableMaterialItem] children, uses a block
+/// layout algorithm, and separates [MaterialSlice] wherever it sees a
+/// [MaterialGap]. Animations work by figuring out where new gaps were
+/// added/removed when the widget gets rebuilt by checking the children's keys,
+/// which is why keys on children are mandatory.
+///
+/// The addition of extra slices is not animated and neither is the adding of
+/// any extra items at the end of the children list.
+class MergeableMaterial extends StatefulWidget {
+  /// Creates a mergeable Material list of items.
+  MergeableMaterial({
+    Key key,
+    this.mainAxis: Axis.vertical,
+    this.elevation: 2,
+    this.children: const <MergeableMaterialItem>[]
+  }) : super(key: key);
+
+  /// The children of the [MergeableMaterial].
+  final List<MergeableMaterialItem> children;
+
+  /// The main layout axis.
+  final Axis mainAxis;
+
+  /// The elevation of all the [Material] slices.
+  final int elevation;
+
+  @override
+  String toString() {
+    return 'MergeableMaterial('
+      'key: $key, mainAxis: $mainAxis, '
+      'elevation: $elevation, children: $children'
+    ')';
+  }
+
+  @override
+  _MergeableMaterialState createState() {
+    return new _MergeableMaterialState();
+  }
+}
+
+class _AnimationTuple {
+  _AnimationTuple({
+    this.controller,
+    this.startAnimation,
+    this.endAnimation,
+    this.gapAnimation,
+    this.gapStart: 0.0
+  });
+
+  final AnimationController controller;
+  final CurvedAnimation startAnimation;
+  final CurvedAnimation endAnimation;
+  final CurvedAnimation gapAnimation;
+  double gapStart;
+}
+
+class _MergeableMaterialState extends State<MergeableMaterial> {
+  List<MergeableMaterialItem> _children;
+  Map<LocalKey, _AnimationTuple> _animationTuples;
+
+  void _initGap(MaterialGap gap) {
+    final AnimationController controller = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+
+    final CurvedAnimation startAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+    final CurvedAnimation endAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    startAnimation.addListener(_handleTick);
+    endAnimation.addListener(_handleTick);
+
+    final CurvedAnimation gapAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    gapAnimation.addListener(_handleTick);
+
+    _animationTuples[gap.key] = new _AnimationTuple(
+      controller: controller,
+      startAnimation: startAnimation,
+      endAnimation: endAnimation,
+      gapAnimation: gapAnimation
+    );
+  }
+
+  @override
+  void initState() {
```

Comment:
this function should come before _initGap, so that the code is in the order it runs